test/PersistentOrderedMap.prop.test.mo

```diff
@@ -0,0 +1,215 @@
+      let key = nextNat(range); (key, debug_show(key)) } )
+  }
+};
+
+let natMap = Map.MapOps<Nat>(Nat.compare);
+
+func mapGen(samples_number: Nat, size: Nat, range: (Nat, Nat)): Iter.Iter<Map.Map<Nat, Text>> {
+  object {
+    var n = 0;
+    public func next(): ?Map.Map<Nat, Text>  {
+      n += 1;
+      if (n > samples_number) { 
+        null 
+      } else {
+        ?natMap.fromIter<Text>(Random.nextEntries(range, size).vals())
+      }
+    }
+  }
+};
+
+
+func run_all_props(range: (Nat, Nat), size: Nat, map_samples: Nat, query_samples: Nat) {
+  func prop(name: Text, prop: Map.Map<Nat, Text> -> Bool): Suite.Suite {
+    test(name, do {
+        for(map in mapGen(map_samples, size, range)) { 
+          if (not prop(map)) {
+            Debug.print("Property \"" # name # "\" failed");
+            Debug.print("M: " # debug_show(Iter.toArray(Map.entries(map))));
+            assert (false);
+          }
+        };
+        true  
+      }, M.equals(T.bool(true)))
+  };
+  func prop_with_key(name: Text, prop: (Map.Map<Nat, Text>, Nat) -> Bool): Suite.Suite {
+    test(name, do {
+        for(map in mapGen(map_samples, size, range)) {
+          for (_query_ix in Iter.range(0, query_samples-1)) {
+            let key = Random.nextNat(range);
+            if (not prop(map, key)) {
+              Debug.print("Property \"" # name # "\" failed");
+              Debug.print("M: " # debug_show(Iter.toArray(Map.entries(map))));
+              Debug.print("K: " # debug_show(key));
+              assert (false);
+            }
+          }
+        };
+        true
+      }, M.equals(T.bool(true)))
+  };
+  run(
+    suite("Property tests",
+   [ 
+      suite("get & put", [
+        prop_with_key("get(put(m, k, v), k) == ?v", func (m, k) {
+          natMap.get(natMap.put(m, k, "v"), k) == ?"v"
+        }),
+        prop_with_key("get(put(put(m, k, v1), k, v2), k) == ?v2", func (m, k) {
+          let (v1, v2) = ("V1", "V2");
+          natMap.get(natMap.put(natMap.put(m, k, v1), k, v2), k) == v2
+        }),
+      ]),
+      
+      suite("replace", [
+        prop_with_key("replace(m, k, v).0 == put(m, k, v)", func (m, k) {
+          natMap.replace(m, k, "v").0 == natMap.put(m, k, "v")
+        }),
+        prop_with_key("replace(put(m, k, v1), k, v2).1 == ?v1", func (m, k) {
+          natMap.replace(natMap.put(m, k, "v1"), k, "v2").1 == ?"v1"
+        }),
+        prop_with_key("get(m, k) == null ==> replace(m, k, v).1 == null", func (m, k) {
+          if (natMap.get(m, k) == null) {
+            natMap.replace(m, k, "v").1 == null
+          } else { true }
+        }),
+      ]),
+
+      suite("delete", [
+        prop_with_key("get(m, k) == null ==> delete(m, k) == m", func (m, k) {
+          if (natMap.get(m, k) == null) {
+            MapMatcher(m).matches(natMap.delete(m, k))
+          } else { true }
+        }),
+        prop_with_key("delete(put(m, k, v), k) == m", func (m, k) {
```

Comment:
We also may want to test, that `put(remove(m, k).0, k, remove(m, k).1) == m`

Reply:
added